### ISSUE TITLE
Infer Subtypes from Template Literal Types

### ISSUE THREAD

**Issue by johnrom** [microsoft/TypeScript#43907]
# Suggestion

## 🔍 Search Terms

Inference
Template Literal Types
Subtypes

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I'd like to be able to infer a subtype from a template literal type. For example:

```ts
const canExtend: `myString.${number}` extends `${infer Key}.${number}` ? true : false = true;
```

## 📃 Motivating Example

This will allow cross-matching between Values and Paths when traversing an object with dot notation.

## 💻 Use Cases

See line 35, [in this Playground](https://www.typescriptlang.org/play?#code/PTCwCgQAgFQCwKZQA4CcD2AjANggthCIcFAMICGAdpegC5QCWlAZgqlLXOfeVAArc4UPNwDGiAM5QABgBIA3k1bsAygFdMyQQF8AdAspq8mNtulQA7okqN6DKcSjlsqBOQAmATycd8ybNxI2Ay0bM4yeJ4CnPryhsam0sQQtJ7ISADymABWUAC8UPIQUFAYdCq0qNFwAFxQAERltBWoAGrOagj1ANzFTqio5J4t1XUAFPXkA0Mt7did9QCUANoAur3g2hCOU4PeWpxSCAAe9jxS0rszldWx8SaoZhC4PNjY1RJ1AIJvHwA8WWyAD41vkoMtGuhyjdBPUADQNK7DGGceGI6bIqqCXQARnq60cmDU9EuGJG2IMRgeZkYUho9BEyC0OCQtHQMgUSjYUAA0ghPHpKQlHuY2RxBgA3NgSJDSQHLADkSPJnAVq2WONW5io7igAHMEPROEhROgBghRPQ5p1dM9DU43taEJ8oE6ALJiOBMPXVAE5BGkvYquB3KmJIFgyZkypOnrbSBgBOwRCMPD+fAISi0bgMdA2dyG8gMbASZLgE7IM1GtJId2e72+p0SBHVKAnUKUdxSaoZZh-JtAiN5PoSSretvHDtd-iCPolaAMZgzzi0pyUbxmqCUBBS1AI1y0NSoGycbhOKQSjoIOdQAD8a88N7qWGyFvo7cz06bN-nJEXrqvKR7AffdDSPGwqEfcAShgu8Hx-KA6ibCcpykVwPDzbBvDGLlVG9XAnRWVYEJKe9Ww-TsLgUUdUG9QVFBYbkADl22qJ5oNgzj7xYydyMnT9u0EXt+0A5Z7jYVYgRIriAPmBAPVocQG0EES5IkMSw1QVYER42hqikjjOM4upt13aSSjqFR8IQJ1pLqPjUKgABrfl0CXb9DKMuCm2WapiM8ozoBgAAJABJFRVysNhWRTcQLScqBmCLEsEWNVwFTpdkLCGDh2VwjgvSkZgMDwHxQjTAJQigYJQkGbBbQC4zlyECjpzkBjlF5fl6Py3S2KSRqjPvPlvFaqQXM8NzZM6UtBq87jWMEFCBOa4SfJGyTzPm6b5PrSgfRU9b+W0qA+sEAyvMuxCtx3NgttgkzbtQe7rtMtgNkJC1yDUGVbCgUQqHpKATGET0EHcGpCAAKih4ooagAAlC0jwkBgpSw-6813egXzfKQxUPdNm2qhgXLhmR5C3cg8AQOpKeYBhUFHOoaPHbQoBpABaCNpEVShqYQNUoAAH3BBV+ZphUEQVBmmdoNUBqgKHEwrKsOBrRHkaZtGECwmA1HTf4BzBZCxqgVn9r6e81j6JDAOWyioBwxi8P2giryIm9bwQ0WNOFfyjN98S9ygXQw6R0QUZ1vWDdwf4rLdmyryBAOYLttSHenCOzXcP4Lb1BFIIurjpN9iapqbVPA8KF7lhGxgbHL9zANWOo6-5BEw90COo-Rzx9cNwQJFUmb288SSq687RlgAUUnQZLT+JuduJ-OU4QtuCXAaHYeg+GADFKo7DhY+df70NCXVMG8Hvtb7gfcBGoCs3ZXg1EoXMbCmg44GJ4IXMhnvGQfMBZC19uLUB0tZajgVFAVYUBuZQAgZLEWSCJYIF0NA+WitlYQFVqgas6QoCH24B2aoD9nR-HIafI4-FHaQTWEOPo1CiaZykDbTy943rPU8vZQQFDaGOWWPlBOeo46CEnqRKAojxErjNvnVBwdpL3najIhA-U7I3TMo1PhnABFsPBPlPgrgGbHE7mHfKSNRySJgmRExDBjgGIUaLJRc0pHtWMQgUx9ESG0DIfwmhfw56VHIIvKxtAERjGcVuTSRFBzsSuq9J6mj84fXLMcSsBD1ZEJ7H2Y2BRcaWgMR5KRkFEJ9F8f4vRgTb6o3vqfI2gFBxQAAGTmzHPtNJeCMlq1SDkwQCklL7SdCPBACJ8l9EKe+OhX5AJe3gp5BcLsG7uDRgwdwahwg-1mg9VaeSmkGOdp1aoiwELQHQkyDGppKA0SLFmBCZElpm1yaMiQxcho7UGV6fajZAItnOsUq8W1HmcC2o9bRl1wV3R0Vo968Z8GEKQD8d4Q9XlMI4lMwFal5mQRvEszqTBVkSnWZs7AKAh5Pj2Wiw5RjBCnMauctwlzvDXNuUwWgDzmoGJeQOZRzVNHcI3rC1AGwgA)

**Comment by johnrom** [comment]
I was able to kind of work around this limitation by using the string value "0" instead of `${number}`, and then converting that to ${number} later down the road. 

It basically breaks TypeScript and causes a lot of "Type instantiation is excessively deep and possibly infinite". You can see changes here, I marked the comments with CHANGE:

[Playground](https://www.typescriptlang.org/play?#code/PTCwCgQAgFQCwKZQA4CcD2AjANggthCIcFAMICGAdpegC5QCWlAZgqlLXOfeVAArc4UPNwDGiAM5QABgBIA3k1bsAygFdMyQQF8AdAspq8mNtulQA7okqN6DKcSjlsqBOQAmATycd8ybNxI2Ay0bM4yeJ4CnPryhsam0sQQtJ7ISADymABWUAC8UPIQUFAYdCq0qNFwAFxQAERltBWoAGrOagj1ANzFTqio5J4t1XUAFPXkA0Mt7did9QCUANoAur3g2hCO0lODw5XVsfEmqGaMUjT0IshaOEi06DIKSmxQANIInnoGRqfnjw4gwAbmwJEhpFlsssAOR7GaHQQw1bLACMq3MVHcUAA5gh6JwkKJ0AMEKJ6HNOroILh6MCOggAGLkBjYCQAdWsMD8AVCABkQmFsHVKQgALJiOBMHHVAA8UIANDJ4QcqoJjn9EgA+fINFWzBk9baQEjEyig1C0aUcKVSQEABmE5AA1ggpCFLCTnTT8VB6fMEOyvRzrPaRQyJbRxNK5YrldNVUd7dIdQVJgmDQGjeAffQtJxI9HKDjRXVqoWpcXRbL0-tMwslVDU3qM4iYqjs4QAFRd4pdqAAJTJalQEgYoOw3jNFvoWGyZNodqetDU-jdSuCrr7MnkUEo5DwCDqu+YDFHtDqEkq1u0UHOAFoddJYfvD8ioAAfKAvg8IGFKmFT3PZEknAKAuzAcAEAAD2QEkCTSJAh1EEcxwnTwYFXXBqgkWVRQkZt8KgGDQkodwpDGV5VGlXBRRWVY+hKAB+RiSigaBSAACQAQQAOQAcQAUTqCw-2wbAnHcbFeAkdJRAYcJRHIcEoGYEk901UclQkJ5RKgJSbFcTA1FZbFCTwSwQiEe0JFYkov2Wep7XqBiwLY+zvyc+olV0XzkNQ8cEEnTC1xw2UVBohBRS1Vz3I85YTjYWK4ocxLUB8vzh1HQLgqwhAwoi4taIZGLWPDAMpBIhAyKkZCSXcWUr1QaUlSoTwtTsqAWLcuLP2-V1PHQZgoHw5Leq-Ioet6tjlk+bwmCgAahpGhkJFWOpZq+DLdH87L0JC7DBFw-DNs8VZSqm3rtGWQToMqchyVlJbhvw7Tr2LC7eo29ZjWAHtt0ZXlSI4PKpFEVxAmxTBvF2tCgowvK5vdShAV4NRKAYdAbGW-M4AkDcGFdGpt2fGFXz-VY+p-N8AKAq8YSgSnHygMnfwZr9WcPXQ6doGFQPAyCOJ4gThKgbjxPQCxcXxK1iygLHJxtJAvL9MFMcoKRlvQEd+COv0FJW4J3HUTBUnSTXCVQCx7AhXHdGTC5LPcaqEDMp5pDt34EjOUCYLgy0OEQqBAe4UjqgOt1ZXD0HiLu6ryKcShPDWJU5gYY2NDNt1Y9IhOmutL80t1fO5cLzTUz6aO10quOau-Maur3BALTK3XOAjmvc6kZYqKgQqcUOzgG+YvvIuqHP46kNOM9NxDbMutimOeeR+8HuAzE6ko6koZu2E6stBA7ie657lg3j4VxT2g7be6HK9h8X-hL4YaDj4T6eTaz+fpsfuR5AvhAV8fjyBDrQMOh9QayluvdR6d9aCp2cOnT+c81hai1BvBe7lt671QJvKAl53o4g2BAP28FA7pDbnADIzA8KrWbHOBcb8p6rT6EvNq+C+igPAe3SBsMcoI1CkdWhFU0FQAAGRQBLkQ36JAuJ8SEnUAAqipQkDRnKq2yljRO2IRCukkSOJAODBo70WjQCw7o7BSGUmOHE+57gcCeNUahwjOgERIbBMhWc+6EIyJQScTiaH4XoTkRhVU674VYYnbwxM3JcOqlXXAuE+H7TymFIJb1mofXEZQ5xQTiGQH+mBfsjIGBBWxDAOefZIKkIDl40UFYYyCBceuShTCcmBLoRXNyUi2nVFYtABgw1x72CifLdgO8LRKlcCuVANhODcCcFIf0nRWJsKTq3Bh5I2kRMwexEggzDauMdm1KZ+IRw2DangtZng8HlSOWEhOEN3AK28JRM+1EipRQZPRPBI9x4PKkH-NKwDe68RItUDBP93JLzBXdf5tcE4BOaRIBKmlzq-LikvepkpGmcGRai72qwlSwtoNUDquyvpNxbhStidRV5fIDBig+nA2nPUOW6DFI8TrVAfpS+FXdl69zmiC95UASUQv5lC6FHwvisq+MtHZUreowvBYIXpghcmrVOuimlP8sURhxcWOUJ05pErFaqzg5KlXTWwdS61WCqV711bSx1qB8k1IQhQoc8QwihG5HgfwgQo5qoBZIwhzYemhr6W5a5fRmVCFDdIe0sRQUWvXpKkeQLNLAO9UYX1CB-WBtCLKcVgh0EZvwa0xNLxRUAKASm0VpbOCQsxcvOtL8c3VTzYMP1PIg1NrgOW-eVaEWAuTFc5ewKK0OuqO6jxtSg7lkNSWBkzSEEiPyH0TZ9BQ2KsfpcqaAzRVMHcOOdOahwi42-i6lo0pfH+I1R0jdoa3nKEoYsTq0AIa3EVmaJqLIUaftkcLBRvgrxqPqCgNVWIlY2GnGwOwBIng+uaqIKDnBOpL35ZPbxmScT3qiI+5FVrpr6oDA0o1TTXqUJ1LuhkGKl65sPD2gtfbi1kuyUioJTLXW3N4wvW1bBuhAA)

**Comment by johnrom** [comment]
For a real world example, here is Formik with the above integration, with strongly typed Fields which enforce valid `name`, `value`, child component, or parse / format function. There are a lot of variants on the core Field component, so there are a lot of ways to test it.

https://codesandbox.io/s/serene-napier-4wjxs?file=/src/App.tsx

**Comment by DanielRosenwasser** [comment]
This seems to work in nightlies, right?

**Comment by DanielRosenwasser** [comment]
Check it out using a nightly release of TypeScript: https://www.typescriptlang.org/play?ts=4.3.0-dev.20210430#code/PTCwCgQAgFQCwKZQA4CcD2AjANggthCIcFAMICGAdpegC5QCWlAZgqlLXOfeVAArc4UPNwDGiAM5QABgBIA3k1bsAygFdMyQQF8AdAspq8mNtulQA7okqN6DKcSjlsqBOQAmATycd8ybNxI2Ay0bM4yeJ4CnPryhsam0sQQtJ7ISADymABWUAC8UPIQUFAYdCq0qNFwAFxQAERltBWoAGrOagj1ANzFTqio5J4t1XUAFPXkA0Mt7did9QCUANoAur3g2hCOU4PeWpxSCAAe9jxS0rszldWx8SaoZhC4PNjY1RJ1AIJvHwA8WWyAD41vkoMtGuhyjdBPUADQNK7DGGceGI6bIqqCXQARnq60cmDU9EuGJG2IMRgeZkYUho9BEyC0OCQtHQMgUSjYUAA0ghPHpKQlHuY2RxBgA3NgSJDSQHLADkSPJnAVq2WONW5io7igAHMEPROEhROgBghRPQ5p1dM9DU43taEJ8oE6ALJiOBMPXVAE5BGkvYquB3KmJIFgyZkypOnrbSBgBOwRCMPD+fAISi0bgMdA2dyG8gMbASZLgE7IM1GtJId2e72+p0SBHVKAnUKUdxSaoZZh-JtAiN5PoSSretvHDtd-iCPolaAMZgzzi0pyUbxmqCUBBS1AI1y0NSoGycbhOKQSjoIOdQAD8a88N7qWGyFvo7cz06bN-nJEXrqvKR7AffdDSPGwqEfcAShgu8Hx-KA6ibCcpykVwPDzbBvDGLlVG9XAnRWVYEJKe9Ww-TsLgUUdUG9QVFBYbkADl22qJ5oNgzj7xYydyMnT9u0EXt+0A5Z7jYVYgRIriAPmBAPVocQG0EES5IkMSw1QVYER42hqikjjOM4upt13aSSjqFR8IQJ1pLqPjUKgABrfl0CXb9DKMuCm2WapiM8ozoBgAAJABJFRVysNhWRTcQLScqBmCLEsEWNVwFTpdkLCGDh2VwjgvSkZgMDwHxQjTAJQigYJQkGbBbQC4zlyECjpzkBjlF5fl6Py3S2KSRqjPvPlvFaqQXM8NzZM6UtBq87jWMEFCBOa4SfJGyTzPm6b5PrSgfRU9b+W0qA+sEAyvMuxCtx3NgttgkzbtQe7rtMtgNkJC1yDUGVbCgUQqHpKATGET0EHcGpCAAKih4ooagAAlC0jwkBgpSw-6813egXzfKQxUPdNm2qhgXLhmR5C3cg8AQOpKeYBhUFHOoaPHbQoBpABaCNpEVShqYQNUoAAH3BBV+ZphUEQVBmmdoNUBqgKHEwrKsOBrRHkaZtGECwmA1HTf4BzBZCxqgVn9r6e81j6JDAOWyioBwxi8P2giryIm9bwQ0WNOFfyjN98S9ygXQw6R0QUZ1vWDdwf4rLdmyryBAOYLttSHenCOzXcP4Lb1BFIIurjpN9iapqbVPA8KF7lhGxgbHL9zANWOo6-5BEw90COo-Rzx9cNwQJFUmb288SSq687RlgAUUnQZLT+JuduJ-OU4QtuCXAaHYeg+GADFKo7DhY+df70NCXVMG8Hvtb7gfcBGoCs3ZXg1EoXMbCmg44GJ4IXMhnvGQfMBZC19uLUB0tZajgVFAVYUBuZQAgZLEWSCJYIF0NA+WitlYQFVqgas6QoCH24B2aoD9nR-HIafI4-FHaQTWEOPo1CiaZykDbTy943rPU8vZQQFDaGOWWPlBOeo46CEnqRKAojxErjNvnVBwdpL3najIhA-U7I3TMo1PhnABFsPBPlPgrgGbHE7mHfKSNRySJgmRExDBjgGIUaLJRc0pHtWMQgUx9ESG0DIfwmhfw56VHIIvKxtAERjGcVuTSRFBzsSuq9J6mj84fXLMcSsBD1ZEJ7H2Y2BRcaWgMR5KRkFEJ9F8f4vRgTb6o3vqfI2gFBxQAAGTmzHPtNJeCMlq1SDkwQCklL7SdCPBACJ8l9EKe+OhX5AJe3gp5BcLsG7uDRgwdwahwg-1mg9VaeSmkGOdp1aoiwELQHQkyDGppKA0SLFmBCZElpm1yaMiQxcho7UGV6fajZAItnOsUq8W1HmcC2o9bRl1wV3R0Vo968Z8GEKQD8d4Q9XlMI4lMwFal5mQRvEszqTBVkSnWZs7AKAh5Pj2Wiw5RjBCnMauctwlzvDXNuUwWgDzmoGJeQOZRzVNHcI3rC1AGwgA

**Comment by DanielRosenwasser** [comment]
(Fixed by #43361, duplicate of #43243)
